### PR TITLE
Standardize on error types for distributed errors.

### PR DESCRIPTION
**Pull request by pritamdamania87**
We have a plethora of error types for various errors raised from c10d. These include `RuntimeError`, `TimeoutError`, `SocketError`, `DistBackendError` etc.

This results in messy code during error handling somewhat like this:
```
if "NCCL" in exception_str:
  ...
if "Timed out initializing process group in store based barrier on rank" in exception_str:
  ...
if "The client socket has timed out after" in exception_str:
  ...
if "Broken pipe" in exception_str:
  ...
if "Connection reset by peer" in exception_str:
  ...
```

To address this issue, in this PR I've ensured added these error types:

1. **DistError** - the base type of all distributed errors
2. **DistBackendError** - this already existed and referred to PG backend errors
3. **DistStoreError** - for errors originating from the store
4. **DistNetworkError** - for general network errors coming from the socket library

### CODE REVIEW

**Review by fduwjj — COMMENTED** [review]
This is great!! Thanks for doing this. I guess there are some test failure, maybe you want to take a look?

**Review by H-Huang — COMMENTED** [review]
Overall changes to consolidate look great to me, but we may want to change the store errors to a name other than `DistBackendError`

If I understand the use case correctly, users need a way to distinguish that the error being thrown is coming from a PyTorch Distributed component to be able to handle it. Consolidating around a specific error type helps with that. This sort of the reason I had added `DistBackendError` (https://github.com/pytorch/pytorch/pull/88134) since that was needed internally and they needed to be able to log/handle the failures coming from PgNCCL.

It might be confusing that the Store will throw `DistBackendError` on things like timeout and socket error, especially since Store can be created on its own irrespective of a comm backend. Would it make sense to have a PyTorch distributed umbrella error type, and then have `DistStoreError` and `DistBackendError`?

**Comment by pritamdamania87** [comment]
> Would it make sense to have a PyTorch distributed umbrella error type, and then have DistStoreError and DistBackendError?

Makes sense, let me add a base error type and then have specializations like this. It actually makes it easier to even distinguish between PG errors and Store errors.

File: torch/csrc/distributed/c10d/Utils.hpp [hunk 1]
```diff
@@ -521,30 +522,30 @@ using SizeType = uint64_t;
         continue;                                                         \
       } else if (                                                         \
           errno_local == WSAETIMEDOUT || errno_local == WSAEWOULDBLOCK) { \
-        TORCH_CHECK(false, "Socket Timeout");                             \
+        C10_THROW_ERROR(DistNetworkError, "Socket Timeout");              \
```
Comment (H-Huang) [inline torch/csrc/distributed/c10d/Utils.hpp:525]: any reason to use `C10_THROW_ERROR` over `TORCH_CHECK`?

File: torch/csrc/distributed/c10d/exception.h [hunk 2]
```diff
@@ -9,37 +9,22 @@
...
+#include <c10/util/Exception.h>
 
-namespace c10d {
-
-class TORCH_API C10dError : public std::runtime_error {
- public:
-  using std::runtime_error::runtime_error;
-
-  C10dError(const C10dError&) = default;
+#define C10D_THROW_ERROR(err_type, msg) \
```
Comment (H-Huang) [inline torch/csrc/distributed/c10d/exception.h:14]: whats the guidance on when to use `C10_THROW_ERROR` vs `C10D_THROW_ERROR` as in some parts I see it is intermingled. Could this be written as a comment?

**Review by H-Huang — COMMENTED** [review]
Epic PR, thanks for big overhaul of the distributed errors @pritamdamania87. Looks good, I will sanity check that there isn't hardcoded dependencies on error types internally.

File: torch/csrc/distributed/c10d/Utils.hpp (hunk 1 above)
Comment (pritamdamania87) [reply torch/csrc/distributed/c10d/Utils.hpp:525]: I feel its simpler with `C10_THROW_ERROR`, `TORCH_CHECK(false,...)` is a bit weird due to the `false` parameter. Also, doesn't `TORCH_CHECK(` always throw `c10:Error`? We need to throw `DistNetworkError` here.

File: torch/csrc/distributed/c10d/exception.h (hunk 2 above)
Comment (pritamdamania87) [reply torch/csrc/distributed/c10d/exception.h:14]: Sure, will add a comment. I guess `C10D_THROW_ERROR` is for c10d:: error types and `C10_THROW_ERROR` is for c10:: error types.

**Comment by pritamdamania87** [comment]
@H-Huang @fduwjj Thanks for the feedback, the PR should be ready for review now. I think I've fixed most of the test failures, will wait for CI to run its course to confirm.

**Review by H-Huang — APPROVED** [review]
Looks great! Internal tests are also good.

**Comment by pritamdamania87** [comment]
@H-Huang I do see messages from "pytorchmergebot", but don't see an explicit merge bot command. I'm assuming you've kicked off the merge process and I don't need to "pytorchbot merge" this PR right?

**Comment by H-Huang** [comment]
oh, my bad, i didnt kick off the merge process. I'll do it now

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting your change, but it is failing inductor test in trunk for one of its model moco' -c nosignals

```
2023-08-28T23:31:52.4840189Z Traceback (most recent call last):
2023-08-28T23:31:52.4840527Z   File "benchmarks/dynamo/torchbench.py", line 481, in <module>
2023-08-28T23:31:52.4841440Z     torchbench_main()
2023-08-28T23:31:52.4841783Z   File "benchmarks/dynamo/torchbench.py", line 477, in torchbench_main
2023-08-28T23:31:52.4842152Z     main(TorchBenchmarkRunner(), original_dir)
2023-08-28T23:31:52.4842510Z   File "/var/lib/jenkins/workspace/benchmarks/dynamo/common.py", line 3026, in main
2023-08-28T23:31:52.4845924Z     process_entry(0, runner, original_dir, args)
2023-08-28T23:31:52.4846295Z   File "/var/lib/jenkins/workspace/benchmarks/dynamo/common.py", line 2983, in process_entry
2023-08-28T23:31:52.4849359Z     return maybe_fresh_cache(
2023-08-28T23:31:52.4849691Z   File "/var/lib/jenkins/workspace/benchmarks/dynamo/common.py", line 1654, in inner
2023-08-28T23:31:52.4851400Z     return fn(*args, **kwargs)
2023-08-28T23:31:52.4851735Z   File "/var/lib/jenkins/workspace/benchmarks/dynamo/common.py", line 3427, in run
2023-08-28T23:31:52.4855527Z     ) = runner.load_model(
2023-08-28T23:31:52.4855829Z   File "benchmarks/dynamo/torchbench.py", line 366, in load_model
2023-08-28T23:31:52.4856236Z     benchmark = benchmark_cls(
2023-08-28T23:31:52.4856655Z   File "/var/lib/jenkins/workspace/torchbench/torchbenchmark/util/model.py", line 25, in __call__
2023-08-28T23:31:52.4857016Z     obj = type.__call__(cls, *args, **kwargs)
2023-08-28T23:31:52.4857376Z   File "/var/lib/jenkins/workspace/torchbench/torchbenchmark/models/moco/__init__.py", line 56, in __init__
2023-08-28T23:31:52.4858007Z     dist.init_process_group(backend='nccl', init_method='tcp://localhost:10001',
2023-08-28T23:31:52.4858548Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/distributed/c10d_logger.py", line 82, in wrapper
2023-08-28T23:31:52.4902859Z     func_return = func(*args, **kwargs)
2023-08-28T23:31:52.4903393Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 1149, in init_process_group
2023-08-28T23:31:52.4951078Z     default_pg, _ = _new_process_group_helper(
2023-08-28T23:31:52.4951631Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 1269, in _new_process_group_helper
2023-08-28T23:31:52.4953078Z     raise ValueError("Distributed package doesn't have NCCL built in")
2023-08-28T23:31:52.4953724Z ValueError: Distributed package doesn't have NCCL built in
2023-08-28T23:31:52.9811240Z ERROR
```

**Pull request closed without merging**